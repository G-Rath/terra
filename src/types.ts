import { AwsResourceType } from '@src/utils';

export enum TFNodeType {
  Resource = 'resource',
  Module = 'module',
  Dynamic = 'dynamic',
  Argument = 'argument',
  Function = 'function',
  Block = 'block',
  Map = 'map'
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface TFBaseNode {
  //
}

export type TFPrimitiveLiteral =
  | string //
  | number
  | boolean
  | null;

export interface TFMapLiteral extends TFBaseNode {
  type: TFNodeType.Map;
  attributes: Array<[string, TFLiteralExpression]>;
}

export type TFLiteralExpression =
  | TFPrimitiveLiteral // todo: replace w/ Node
  | TFFunctionCall
  | TFLiteralExpression[] // todo: replace w/ Node
  | TFMapLiteral;

interface TFFunctionCall extends TFBaseNode {
  type: TFNodeType.Function;
  name: string;
  args: unknown[];
}

export interface TFArgument<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Argument;
  identifier: TIdentifier;
  expression: TFLiteralExpression;
}

export interface TFBlockLiteral<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Block;

  /**
   * The name of the block literal
   */
  name: string;
  /**
   * The arguments that make up the resource.
   */
  body: TFBlockBody<TIdentifier>;
}

export interface TFDynamicBlock<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Dynamic;
  /**
   * The label of the dynamic block, which specifies what kind of nested block to generate.
   */
  name: string;
  content: TFBlockBody<TIdentifier>;
  labels: string[];
  forEach: unknown;
  /**
   * Sets the name of a temporary variable that represents the current element of the complex value.
   * If omitted, the name of the variable defaults to the label of the dynamic block.
   */
  iterator?: string;
}

export type TFBlockBodyElement<TIdentifier extends string = string> =
  | TFArgument<TIdentifier>
  | TFBlockLiteral
  | TFDynamicBlock;

export type TFBlockBody<TIdentifier extends string = string> = Array<
  TFBlockBodyElement<TIdentifier>
>;

export enum RandomResourceType {
  RandomString = 'random_string',
  RandomPet = 'random_pet',
  RandomId = 'random_id'
}

export type ResourceType =
  | AwsResourceType //
  | RandomResourceType
  | string;

export type TFTopLevelBlock = TFResourceBlock | TFModuleBlock;

/** The AST for a valid Terraform file. */
export type TFFileAST = TFTopLevelBlock[];

/**
 * A Terraform module, made up of ASTs of the files in the module.
 * Note that a "module" is different from a "module block".
 */
interface TFModule {
  files: { [K: string]: TFFileAST };
}

export interface TFResourceBlock<TIdentifier extends string = string> {
  type: TFNodeType.Resource;
  /**
   * The type of this resource.
   */
  resource: ResourceType;
  /**
   * The name of this resource.
   */
  name: string;
  /**
   * The body of this resource block.
   */
  body: TFBlockBody<TIdentifier>;
}

export interface TFModuleBlock<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Module;
  /**
   * The name of this module.
   */
  name: string;
  /**
   * The body of this module block.
   */
  body: TFBlockBody<TIdentifier | 'source'>;
}
