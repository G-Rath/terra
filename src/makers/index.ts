export * from './makeTFLabel';
export * from './makeTFBlock';
export * from './makeTFAttribute';
export * from './makeTFAttributeKey';
export * from './makeTFArgument';
export * from './makeTFBlockBody';
export * from './makeTFIdentifier';
export * from './makeTFMapExpression';
export * from './makeTFModuleBlock';
export * from './makeTFDynamicBlock';
export * from './makeTFFunctionExpression';
export * from './makeTFResourceBlock';
export * from './makeTFSimpleLiteral';
export * from './makeTFStringArgument';
export * from './makeTFListExpression';
