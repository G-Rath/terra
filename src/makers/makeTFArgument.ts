import { TFArgument, TFLiteralExpression, TFNodeType } from '@src/types';

export const makeTFArgument = <TIdentifier extends string = string>(
  identifier: TIdentifier,
  expression: TFLiteralExpression,
  surroundingText?: Partial<TFArgument['surroundingText']>
): TFArgument<TIdentifier> => ({
  type: TFNodeType.Argument,
  identifier,
  expression,
  surroundingText: {
    leadingInnerText: '',
    trailingInnerText: '',
    ...surroundingText
  }
});
