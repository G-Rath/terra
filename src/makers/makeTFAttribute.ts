import {
  makeTFListExpression,
  makeTFAttributeKey,
  makeTFSimpleLiteral
} from '@src/makers';
import {
  TFIdentifier,
  TFLabel,
  TFLiteralExpression,
  TFNodeType,
  TFAttribute,
  TFAttributeValue
} from '@src/types';

const buildValue = (
  value: TFLiteralExpression | string | Array<TFLiteralExpression | string>
): TFAttributeValue => {
  if (typeof value === 'string') {
    return makeTFSimpleLiteral(value);
  }

  if (Array.isArray(value)) {
    return makeTFListExpression(value);
  }

  return value;
};

export const makeTFAttribute = <TKey extends string = string>(
  key: TKey | TFIdentifier<TKey> | TFLabel<TKey>,
  value: TFAttributeValue | string | Array<TFAttributeValue | string>,
  surroundingText?: Partial<TFAttribute['surroundingText']>
): TFAttribute<TKey> => ({
  type: TFNodeType.Attribute,
  key:
    typeof key === 'string' //
      ? makeTFAttributeKey(key)
      : key,
  value: buildValue(value),
  surroundingText: {
    leadingInnerText: '',
    trailingInnerText: '',
    ...surroundingText
  }
});
