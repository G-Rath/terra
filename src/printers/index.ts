export * from './printBlockBody';
export * from './printBlockLiteral';
export * from './printFileAST';
export * from './printResourceBlock';
