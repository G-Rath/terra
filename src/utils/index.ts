export * from './AwsResourceType';
export * from './asResourceName';
export * from './makeTFArgument';
export * from './makeTFBlockLiteral';
export * from './makeTFDynamicBlock';
export * from './makeTFResourceBlock';
export * from './makeTFStringArgument';
export * from './normaliseRoute53Name';
