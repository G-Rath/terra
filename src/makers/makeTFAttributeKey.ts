import { makeTFIdentifier, makeTFLabel } from '@src/makers';
import { TFAttributeKey } from '@src/types';

export const makeTFAttributeKey = <TKey extends string = string>(
  key: TKey,
  surroundingText?: Partial<TFAttributeKey['surroundingText']>
): TFAttributeKey<TKey> => {
  if (['"', "'"].includes(key[0])) {
    return makeTFLabel(key, surroundingText);
  }

  return makeTFIdentifier(key, surroundingText);
};
