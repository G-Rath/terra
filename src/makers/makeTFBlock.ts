import { makeTFBlockBody } from '@src/makers/makeTFBlockBody';
import { makeTFLabel } from '@src/makers/makeTFLabel';
import {
  TFBlock,
  TFBlockBody,
  TFBlockBodyBody,
  TFLabel,
  TFNodeType,
  TFResourceBlock
} from '@src/types';

export const makeTFBlock = <TIdentifier extends string = string>(
  blockType: string,
  labels: Array<TFLabel | string>,
  body: TFBlockBody<TIdentifier> | TFBlockBodyBody<TIdentifier>,
  surroundingText?: Partial<TFResourceBlock['surroundingText']>
): TFBlock<TIdentifier> => ({
  type: TFNodeType.Block,
  blockType,
  labels: labels.map(label =>
    typeof label === 'string' ? makeTFLabel(label) : label
  ),
  body: Array.isArray(body) ? makeTFBlockBody(body) : body,
  surroundingText: {
    leadingOuterText: '',
    trailingOuterText: '',
    ...surroundingText
  }
});
