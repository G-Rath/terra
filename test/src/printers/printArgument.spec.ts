import { makeTFArgument } from '@src/makers';
import { printArgument, printLiteralExpression } from '@src/printers';
import { TFNodeType } from '@src/types';
import { mocked } from 'ts-jest/utils';

jest.mock('@src/printers/printLiteralExpression');

describe('printArgument', () => {
  beforeEach(() =>
    mocked(printLiteralExpression).mockReturnValue(
      nameof(printLiteralExpression)
    )
  );

  it('prints the expression using printLiteralExpression', () => {
    expect(printArgument(makeTFArgument('name', '"world"'))).toMatchSnapshot();
  });

  it.todo('quotes the identifier when required');

  describe('when the expression has braces', () => {
    beforeEach(() =>
      mocked(printLiteralExpression).mockReturnValue(
        ['{', `  ${nameof(printLiteralExpression)}`, '}'].join('\n')
      )
    );

    it('prints the opening brace on the first line', () => {
      expect(
        printArgument(
          makeTFArgument('name', {
            type: TFNodeType.Map,
            attributes: [
              [
                'MyArray',
                [
                  'aws_subnet.public_a.id',
                  'aws_subnet.public_b.id',
                  'aws_subnet.public_c.id'
                ]
              ]
            ]
          })
        )
      ).toMatchSnapshot();
    });
  });
});
