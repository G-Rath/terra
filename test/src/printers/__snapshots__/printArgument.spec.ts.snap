// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printArgument prints the expression using printLiteralExpression 1`] = `"name = printLiteralExpression"`;

exports[`printArgument when the expression has braces prints the opening brace on the first line 1`] = `
"name = {
  printLiteralExpression
}"
`;
